csharp/src/Apache.Arrow/C/CArrowSchemaExporter.cs

```diff
@@ -0,0 +1,265 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Apache.Arrow.Types;
+
+namespace Apache.Arrow.C
+{
+    public static class CArrowSchemaExporter
+    {
+        /// <summary>
+        /// Export a type to a <see cref="CArrowSchema"/>.
+        /// </summary>
+        /// <param name="datatype">The datatype to export</param>
+        /// <param name="schema">An allocated but uninitialized CArrowSchema pointer.</param>
+        /// <example>
+        /// <code>
+        /// CArrowSchema* exportPtr = CArrowSchema.Create();
+        /// CArrowSchemaExporter.ExportType(dataType, exportPtr);
+        /// foreign_import_function(exportPtr);
+        /// CArrowSchema.Free(exportPtr);
```

Comment:
Won't this trigger a call to `exportPtr->release`?  Is that safe given that you've already passed ownership during the `foreign_import_function`.

Reply:
Okay read up on it. In the C++ bridge, the consumer moves the data a struct it owns by memcpy'ing it to the struct and then calling release on the source. So all import functions are expected to call `release` on the original struct (but might copy data off of it before then).

Reply:
https://github.com/apache/arrow/blob/427f1b649e36ef3c64dd61ed86808463a440eb94/cpp/src/arrow/c/helpers.h#L70-L75

Reply:
So let's split these into two points.

1. I agree that `CArrowSchema*` needs to be freed.  I'm not sure it has to be heap-allocated.  I think it could be stack-allocated but I don't remember if `struct` was guaranteed to be stack allocated or not.  It's probably large enough (with all the strings) to justify heap allocation so I think this is good.

2. I'm not sure that it is safe to assume that `release` has been set to null.  I do agree it is safe when calling Arrow-C++'s import function.  I just don't think this is part of the spec.  And, the spec is very clear that it is the consumer's responsibility to call the top-level `release` function.  So, while Arrow does copy everything into its own memory during it's import function I don't think this is spec-mandated.  Maybe it would be a good idea to add this mandate to the spec?

Reply:
Ah, regarding point 1, the strings are just pointers in the struct so...if you could mandate it is stack-allocated, I think that would be fine too.  But still, this is unlikely to really matter.  These structs/functions should be more or less invisible to the user most of the time.

Reply:
I was thinking that stack-allocated values in C# would be managed memory, and that wouldn't be accessible from C/other languages. But I might be mistaken on that.

Reply:
> I'm not sure that it is safe to assume that release has been set to null. I do agree it is safe when calling Arrow-C++'s import function. I just don't think this is part of the spec

You're right this isn't guaranteed in the spec. It's more something that needs to be negotiated in each API. So perhaps I just need to say something like that in the examples?

Reply:
Well, the problem is that you really DO need to release the `CArrowSchema*` regardless of what the foreign import function does.  So, is there a way to do that without calling `release` if it exists?

Maybe the pattern could be:

```
/// CArrowSchema* exportPtr = CArrowSchema.Create();
/// CArrowSchemaExporter.ExportType(dataType, exportPtr);
/// foreign_import_function(exportPtr);
/// exportPtr->release = IntPtr.Zero;
/// CArrowSchema.Free(exportPtr);
```

However, this seems potentially error prone.

Reply:
I think there are two hard rules:

 * `CArrowSchema` should only be free'd if `CArrowSchema.release` is null.
 * `CArrowSchema.release` should only be set to null when either the release callback is called or the data is moved. (So I think that line you added is not ideal)

If the `foreign_import_function` doesn't set it null, then a different pattern should be used. For example, the foreign importer should be the one to allocate the CArrowSchema and be responsible for freeing it. I think this is the reason that it's intended that the consumers allocate and own `CArrowSchema`, rather than the producer. But when interacting with R and Python, it's sometimes convenient to have the C++/Rust/... producer make these structs, so the move pattern is nice.

Reply:
I was all set to disagree with you. However, reading in detail, I think you're correct.

I don't like this but it might be the safest general pattern:

```
/// CArrowSchema* exportPtr = CArrowSchema.Create();
/// CArrowSchemaExporter.ExportType(dataType, exportPtr);
/// foreign_import_function(exportPtr);
/// if (exportPtr->release != IntPtr.Zero)
/// {
///     // Generally a foreign import function should copy the release callback
///     // and set this to IntPtr.Zero.  If they do not then it will not be possible for
///     // them to call the release callback.  Perhaps the imported expected to allocate
///     // the CArrowSchema instance itself?
///     CArrowSchema.Free(exportPtr);
///     throw new InvalidOperationException("Foreign import function failed to fully consume schema");
/// }
/// CArrowSchema.Free(exportPtr);
```

Reply:
I think this discussion is a bit pedantic as it all depends how the consumer API (`foreign_import_function`) is defined, and that is not specified by the C Data Interface.

The C Data Interface only says what should happen with the _contents_ of the C struct. How the struct itself is allocated is left for the producer or consumer to decide (and may indeed depend on which language those are implemented in).

I would generally suggest that the producer handles the lifetime of the struct it populates, and the consumer moves the contents of the producer-populated struct to its own managed area.

(also, in real-world examples you would hope that `foreign_import_function` is able to return an error)